query-algebrizer/src/types.rs

```diff
@@ -194,6 +194,88 @@ pub enum ColumnConstraint {
     HasType(TableAlias, ValueType),
 }
 
+#[derive(PartialEq, Eq, Debug)]
+pub enum ColumnConstraintOrAlternation {
+    Constraint(ColumnConstraint),
+    Alternation(ColumnAlternation),
+}
+
+impl From<ColumnConstraint> for ColumnConstraintOrAlternation {
+    fn from(thing: ColumnConstraint) -> Self {
+        ColumnConstraintOrAlternation::Constraint(thing)
+    }
+}
+
+/// An ordered collection of constraints that must all match.
+#[derive(PartialEq, Eq)]
+pub struct ColumnIntersection(pub Vec<ColumnConstraintOrAlternation>);
+
+impl From<Vec<ColumnConstraint>> for ColumnIntersection {
+    fn from(thing: Vec<ColumnConstraint>) -> Self {
+        ColumnIntersection(thing.into_iter().map(|x| x.into()).collect())
+    }
+}
+
+impl Default for ColumnIntersection {
+    fn default() -> Self {
+        ColumnIntersection(vec![])
+    }
+}
+
+impl IntoIterator for ColumnIntersection {
+    type Item = ColumnConstraintOrAlternation;
+    type IntoIter = ::std::vec::IntoIter<ColumnConstraintOrAlternation>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl ColumnIntersection {
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn also(&mut self, constraint: ColumnConstraint) {
+        self.0.push(ColumnConstraintOrAlternation::Constraint(constraint));
+    }
+}
+
+/// An ordered collection of a collection of constraints, one of which must entirely match.
+#[derive(PartialEq, Eq, Debug)]
+pub struct ColumnAlternation(pub Vec<ColumnIntersection>);
+
+impl Default for ColumnAlternation {
+    fn default() -> Self {
+        ColumnAlternation(vec![])
+    }
+}
+
+impl IntoIterator for ColumnAlternation {
+    type Item = ColumnIntersection;
+    type IntoIter = ::std::vec::IntoIter<ColumnIntersection>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl ColumnAlternation {
+    pub fn instead(&mut self, intersection: ColumnIntersection) {
```

Comment:
With `instead` and `also` I was trying to avoid using common control flow words. I think I might opt for `add_alternate` and `add_intersection`…